### PR TITLE
tf: Improvements to traffic generator

### PR DESCRIPTION
**Pull request by howardjohn**
A few refactorings for traffic generator to make its usage cleaner.
There is a lot of boilerplate when you have multiple generators.

### CODE REVIEW

File: pkg/test/framework/components/echo/util/traffic/result.go [hunk 1]
```diff
@@ -49,12 +50,25 @@ func (r *Result) add(resp client.ParsedResponses, err error) {
...
 	}
 }
 
 func (r Result) PercentSuccess() float64 {
-	return (float64(r.SuccessfulRequests) / float64(r.TotalRequests)) * 100.0
+	return float64(r.SuccessfulRequests) / float64(r.TotalRequests)
+}
+
+// CheckSuccessRate asserts that a minimum success threshold was met.
+func (r Result) CheckSuccessRate(t test.Failer, minimumPercent float64) {
```
Comment (Monkeyanator) [inline pkg/test/framework/components/echo/util/traffic/result.go:64]: nit: name doesn't imply this fails test (like how tf `Check...` return error)

File: pkg/test/framework/components/echo/util/traffic/generator.go [hunk 2]
```diff
@@ -65,12 +69,13 @@ var _ Generator = &generator{}
...
 type generator struct {
 	Config
+	t       test.Failer
 	result  Result
 	stop    chan struct{}
 	stopped chan struct{}
 }
 
-func (g *generator) Start() {
+func (g *generator) Start() Generator {
```
Comment (Monkeyanator) [inline pkg/test/framework/components/echo/util/traffic/generator.go:78]: nit: do we want to prevent calling start twice on same generator?

**Review by Monkeyanator — APPROVED** [review]

File: pkg/test/framework/components/echo/util/traffic/result.go (hunk 1 above)
Comment (nmittler) [reply pkg/test/framework/components/echo/util/traffic/result.go:64]: +1 you could just have 2 versions of the method. `CheckSuccessRate` that returns `error` and then `CheckSuccessRateOrFail` that calls `CheckSuccessRate` and fails a `Failer` if there's an error.

Comment (nmittler) [reply pkg/test/framework/components/echo/util/traffic/result.go:64]: We've been following this pattern pretty consistently throughout the test framework. There have been occasions where it's been useful. But I'm fine with leaving it as-is for now. We can always refactor later if it ends up being useful.

**Review by nmittler — APPROVED** [review]

**Pull request merged**